### ISSUE TITLE
Issue of early stopping for sparsity

### ISSUE THREAD

**Issue by jiweiqi** [jiweiqi/CellBox.jl#15]
Start a thread on the potential issue of early stopping on parameter inference (identify sparsity)

Currently, the rule of early stopping that we implemented is to stop training if the validation loss reaches a plateau. This is justified for deep learning with the goal of data fitting. For the deep learning model, the loss landscape is assumed to be flat near (sub) global minima, thus it is unnecessary to further training as long as the point is fell into that good valley.

Here, we additionally want a sparse model. But whether the model is sparse has little effect on the fitness by definition. Therefore, it is very likely early stopping will stop the model while the model is far away from sparse yet.

Instead, we shall train the model for a very long time.

Other tips I am not sure are:
* small weight decay seems to be better than a large one
* learning rate decay seems to be useful to further encourage sparsity, as shown below (or:5e-3 -> 1.e-3)

![image](https://user-images.githubusercontent.com/8445647/108387729-dd545480-71db-11eb-8f7b-9084869f6e44.png)

**Comment by jiweiqi** [comment]
Alternatively, we could use weight pruning to encourage sparsity, although this assumes that all of the important wij has a large absolute value.
Example at
https://github.com/DENG-MIT/CRNN/blob/a94b20604fce305a55854a9e34c45fa2b28de8a8/case1/case1_hardthreshhold.jl#L76
![image](https://user-images.githubusercontent.com/8445647/108388443-a5014600-71dc-11eb-9c9c-7aaa3ea280df.png)

**Comment by DesmondYuan** [comment]
All great! 

I think pruning is a good idea. Does Julia have a similar prune training function as PyTorch/Keras?

**Comment by jiweiqi** [comment]
I don't think there is one in Julia. Normally, I do it manually. Actually, I don't retrain the model after pruning since I use a very tight threshold and the performance is almost unchanged after pruning.

**Comment by DesmondYuan** [comment]
Another observation is that for smaller datasets, too small lr would result in local optimum overfitting.
